File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
    
            prefix=[]
            prefix.append(0)
            for i in range(1,len(prices)):
                    prefix.append(prices[i]-prices[i-1])
        
            total=0
            maxim=0
            for item in prefix:
                    total+=item
                    if total<0:
                        total=0
                    
                    maxim=max(total,maxim) 
            
            
            return maxim
            
                

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        
            int n =prices.size();
            vector<int>prefix;
            prefix.push_back(0);
            for(int i=1;i<prices.size();i++){
                
                    prefix.push_back(prices[i]-prices[i-1]); 
                
            }
            
            int max_sum=0;
            int sum=0;
            for(int j=0;j<prices.size();j++){
                    
                    sum+=prefix[j];
                    if(sum<0){
                            sum=0;
                    }    
                    max_sum=max(max_sum,sum);
            }
            
            return max_sum;
                
        
        
    }
};